### ISSUE TITLE
AVIATOR

### ISSUE THREAD

**Issue by Wilen047** [Wilen047/AVIATOR-PREDICTION-APK#10]
# Issue #10: Feature Engineering for Analytics & Machine Learning

**Status:** Open

**Priority:** P1

**Labels:** `machine-learning`, `feature-engineering`, `data-science`, `backend`

## Description

Develop the feature engineering pipeline that transforms cleaned historical Aviator round data into structured numerical features for statistical analysis and machine-learning experiments.

The features must be generated using only information that would have been available at the relevant historical time. This prevents data leakage and makes model evaluation more meaningful.

The features should be treated as analytical inputs, not as guarantees of future Aviator outcomes.

## Objectives

* Identify useful historical features.
* Create numerical features from round history.
* Create rolling-window features.
* Create frequency-based features.
* Create sequence-based features.
* Normalize/scale features where required.
* Prevent data leakage.
* Store or generate processed features for model training.
* Make the feature pipeline reusable.

### 1. Current Historical Features

For each historical record, generate features such as:

```text
previous_multiplier
previous_2_multiplier
previous_3_multiplier
previous_5_multiplier
previous_10_multiplier
```

These represent previous observations only.

### 2. Rolling Statistics

Generate rolling statistics using configurable windows:

```text
rolling_mean_5
rolling_mean_10
rolling_mean_25
rolling_mean_50
rolling_median_10
rolling_std_10
rolling_std_25
```

Example:

```text
Last 10 rounds
→ calculate mean
→ calculate median
→ calculate standard deviation
```

### 3. Threshold Frequencies

Calculate historical frequencies within a rolling window.

Examples:

```text
count_above_2x
count_above_3x
count_above_5x
count_below_2x
```

These should be calculated from previous rounds only.

### 4. Multiplier Range Features

Create counts for configurable multiplier ranges:

```text
count_1x_to_1_49x
count_1_5x_to_1_99x
count_2x_to_2_99x
count_3x_to_4_99x
count_5x_to_9_99x
count_10x_plus
```

### 5. Sequence Features

Generate historical sequence information.

Examples:

```text
previous_is_low
previous_is_high
low_streak_length
high_streak_length
```

Thresholds must be configurable.

Example:

```text
Low  = < 2.00x
High = >= 2.00x
```

### 6. Time-Based Features

Where appropriate, generate:

```text
hour
day_of_week
day_of_month
```

Avoid assuming that time-of-day features have predictive value unless supported by validation.

## Feature Pipeline

The pipeline should follow:

```text
Clean Historical Data
        ↓
Sort Chronologically
        ↓
Generate Lag Features
        ↓
Generate Rolling Features
        ↓
Generate Frequency Features
        ↓
Generate Sequence Features
        ↓
Generate Time Features
        ↓
Remove Invalid Training Rows
        ↓
Validate Features
        ↓
Save Feature Dataset
```

## Prevent Data Leakage

This is a critical requirement.

Features for a given round must not use:

* Future rounds.
* Future multipliers.
* Information generated after the target event.
* Statistics calculated using the target itself where inappropriate.

Example:

Incorrect:

```text
rolling_mean = current + previous rounds
```

Correct:

```text
rolling_mean = previous rounds only
```

Use appropriate shifting operations before calculating rolling statistics.

## Feature Validation

Check for:

* Missing values.
* Infinite values.
* Incorrect data types.
* Unexpected ranges.
* Duplicate rows.
* Data leakage.
* Incorrect chronological ordering.

## Feature Scaling

Depending on the selected model, support:

* Standardization.
* Min-max scaling.
* Robust scaling.

Scaling parameters must be fitted only on training data and then applied to validation/test data.

## Feature Dataset

The resulting dataset may contain:

```text
played_at
previous_multiplier
rolling_mean_10
rolling_std_10
count_above_2x
count_above_5x
low_streak_length
high_streak_length
hour
day_of_week
```

The exact feature set should be finalized after exploratory analysis.

## Implementation

Use:

* Pandas
* NumPy
* Scikit-learn where appropriate

Suggested structure:

```text
app/
├── ml/
│   ├── features/
│   │   ├── lag_features.py
│   │   ├── rolling_features.py
│   │   ├── frequency_features.py
│   │   ├── sequence_features.py
│   │   └── time_features.py
│   │
│   └── feature_pipeline.py
│
└── schemas/
    └── features.py
```

## API / Service Integration

Feature generation should primarily be implemented as a reusable service or background job rather than running expensive processing directly inside normal user requests.

Possible administrative endpoint:

```text
POST /api/v1/ml/features/generate
```

## Feature Versioning

Each generated feature dataset should have a version.

Example:

```text
feature_version = 1.0
```

Store metadata such as:

```text
feature_version
created_at
source_dataset
feature_names
window_sizes
processing_version
```

This ensures that model experiments can be reproduced.

## Testing Requirements

Create tests for:

* [ ] Lag features.
* [ ] Rolling mean.
* [ ] Rolling median.
* [ ] Rolling standard deviation.
* [ ] Threshold frequencies.
* [ ] Multiplier range counts.
* [ ] Low/high streak calculations.
* [ ] Time-based features.
* [ ] Missing-value handling.
* [ ] Feature scaling.
* [ ] Chronological ordering.
* [ ] Data leakage detection.
* [ ] Feature versioning.

## Acceptance Criteria

* [ ] Feature engineering pipeline is implemented.
* [ ] Lag features are generated.
* [ ] Rolling statistics are generated.
* [ ] Threshold features are generated.
* [ ] Sequence features are generated.
* [ ] Time-based features are generated where appropriate.
* [ ] Data leakage prevention is implemented.
* [ ] Feature validation is implemented.
* [ ] Scaling pipeline is implemented where required.
* [ ] Feature versions are tracked.
* [ ] Feature generation is reproducible.
* [ ] Unit tests pass.
* [ ] Feature dataset is ready for model training.

## Expected Outcome

A reusable and leakage-safe feature engineering pipeline that converts historical Aviator data into structured features suitable for statistical experiments and machine-learning model development.

## Next Issue

**Issue #11: Develop Machine Learning Model for Historical Data Analysis**
